Add mesos containerizer support

sashankreddya opened the pull request:
This commit includes support for collecting container stats
launched by mesos containerizer.

andyxning committed:
add raw container cgroup path prefix whitelist

sashankreddya committed:
Add mesos containerizer support

This commit includes support for collecting container stats
launched by mesos containerizer.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/cadvisor?pullRequest=2) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/uber/cadvisor?pullRequest=2) before we can accept your contribution.<br/>**1** out of **7** committers have signed the CLA.<br/><br/>:white_check_mark: sashankreddya<br/>:x: rohitagarwal003<br/>:x: jessfraz<br/>:x: dashpole<br/>:x: MaximilianMeister<br/>:x: andyxning<br/>:x: moooofly<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/cadvisor?pullRequest=2) it.</sub>

kiranbond commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
```
Does this need to be Google?

sashankreddya replied:
Good question. All the files in this repo have it as "Google", so had it this way. Can check with David on this ?

sashankreddya replied:
A quick grep on the cadvisor directory shows this to be only "Google"
`grep -r "Copyright" * | grep -v "Google" | grep -v "vendor" | less`

Still waiting on reply from David.

sashankreddya replied:
I got a reply from David, suggesting that all the file headers should have the copyright associated with "Google"

cinchurge replied:
[here](https://haacked.com/archive/2006/01/26/WhoOwnstheCopyrightforAnOpenSourceProject.aspx/) is a good read on the topic

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+)
+
+const (
+	slaveURL               = "http://localhost:5051"
```
is there a way to configure this for cadvisor? although uncommon, it is possible for people to use a port different from 5051, or that they're not listening on localhost, but a specific ip address

sashankreddya replied:
We can add command line arguments support if we plan to support custom address and port.

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+	slaveURL               = "http://localhost:5051"
+	containersHTTPEndpoint = "/containers"
+	slaveStateHTTPEndpoint = "/state"
+	timeout                = 10 * time.Second
```
similarly this should be configurable. hardcoding things like this cause lots of headaches in edge cases

sashankreddya replied:
Sure we can have slaveURL and timeout added as config through command line which is the preferred way with cadvisor

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+	url string
+}
+
+type mesosSlaveClient interface {
```
maybe use `mesosAgentClient` instead as "agent" is now the accepted replacement for "slave"

sashankreddya replied:
Sure

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+
+// Client is an interface to query mesos slave http endpoints
+func Client() (mesosSlaveClient, error) {
+	mesosClientOnce.Do(func() {
```
why not just a proper constructor (e.g. something like `func New`)? this doesn't seem like a pattern people commonly use for Go...

sashankreddya replied:
This is the style followed in all the containers already present in Cadvisor, so followed their style here.

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+	ContainerInfo(id string) (*containerInfo, error)
+}
+
+type containerInfo struct {
```
defining structs without a corresponding interface (something supposedly returned by `New`) make it EXTREMELY difficult for downstream consumers to mock dependencies. I believe the more idiomatic way to design this is to define a constructor along the lines of `func New(...) Client { return *client{...} }` and a corresponding `type Client interface{...}`

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+func (self *client) ContainerInfo(id string) (*containerInfo, error) {
+	c, err := self.getContainer(id)
+	if err != nil {
+		return nil, err
```
is `github.com/pkg/errors` allowed in this code base? if yes, maybe wrap the err so we can get stack traces

sashankreddya replied:
I did a search for this package in here and didn't see it being supported.

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+
+	labels, err = fetchLabelsFromExecutor(exec)
+	if err != nil {
+		return nil, fmt.Errorf("error while fetching labels from executor")
```
`error.New` is preferred when the message is just a static string as opposed to a format string

sashankreddya replied:
sure

cinchurge commented on container/mesos/client.go:
```diff
@@ -0,0 +1,158 @@
+		return fmt.Errorf("error creating HTTP request to %s: %s", url, err)
+	}
+
+	res, err := self.hc.Do(req)
```
are we doing retries here?

sashankreddya replied:
Nope. Will add retry support

cinchurge commented on container/mesos/factory.go:
```diff
@@ -0,0 +1,142 @@
+}
+
+// The mesos factory can handle any container.
+func (self *mesosFactory) CanHandleAndAccept(name string) (bool, bool, error) {
```
using named returns will make this self-documenting.. otherwise I'll have to trace the code to figure out what the first and second bools represent. this is especially problematic for exported methods, which this is.

sashankreddya replied:
Sure added it.

cinchurge commented on container/mesos/factory.go:
```diff
@@ -0,0 +1,142 @@
+	client mesosSlaveClient
+}
+
+func (self *mesosFactory) String() string {
```
what does it mean to export methods for a type that isn't exported? code outside of this package will never be able to use the type if it isn't exported, hence will have no way to call the exported methods. On the other hand, if you intend to export the factory, you should export an interface as opposed to a struct.

sashankreddya replied:
This implements the "ContainerHandlerFactory" interface at https://github.com/google/cadvisor/blob/master/container/factory.go#L26

Added a comment to make it more readable.

dashpole committed:
dont watch .mount cgroups

dashpole committed:
Merge pull request #1964 from dashpole/dont_watch_mount

Dont watch "*.mount" cgroups

jessfraz committed:
vendor: update docker import

Signed-off-by: Jess Frazelle <acidburn@microsoft.com>

dashpole committed:
Merge pull request #1967 from jessfraz/vendor

vendor: update docker import

rohitagarwal003 committed:
Initialize NVML on demand.

Earlier if the NVIDIA driver was not installed when cAdvisor was started
we would start a goroutine to try to initialize NVML every minute.

This resulted in a race. We can have a situation where:
- goroutine tries to initialize NVML but fails. So, it sleeps for a minute.
- the driver is installed.
- a container that uses NVIDIA devices is started.
This container would not get GPU stats because a minute has not passed
since the last failed initialization attempt and so NVML is not
initialized.

dashpole committed:
Merge pull request #1969 from mindprince/fix-race

Initialize NVML on demand.

MaximilianMeister committed:
fix kubernetes issue #65204

https://github.com/kubernetes/kubernetes/issues/65204

Signed-off-by: Maximilian Meister <mmeister@suse.de>

dashpole committed:
Merge pull request #1971 from MaximilianMeister/fix-kubernetes-issue-65204

fix brtfs issue where dir is sub-directory of a sub-volume

dashpole committed:
Merge pull request #1926 from andyxning/add_docker-only_whitelist

add docker_only_prefix_whitelist

sashankreddya committed:
Add dependencies needed for mesos containerizer

moooofly committed:
fix: ./deploy/build.sh can not find Dockerfile

moooofly committed:
fix: update docs/deploy.md

dashpole committed:
Merge pull request #1983 from moooofly/develop

fix: ./deploy/build.sh can not find Dockerfile
